### ISSUE TITLE
[ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type Function. Received type object 

### ISSUE THREAD

**Issue by rdrnt** [customerio/customerio-node#79]
I'm following your [documentation](https://github.com/customerio/customerio-node#ciotrackid-data) for tracking events in Node.js. I'm running into an issue where I keep getting

`[ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type Function. Received type object `

Here's my sample code:

```
	const userId = 'adajkdaAS7aAS';
	
	const customerioTrackClient = new TrackClient(
      'mySiteId',
      'myApiKey',
      { region: RegionUS }
    );

      await customerioTrackClient.track(userId, {
        name: 'ExampleName',
        data: {
          sku: 'blahlah',
        },
      });
```

Any ideas?

**Comment by swastik** [comment]
Hi @rdrnt—thanks for the reporting this issue! Can you please tell me the version of `node` and `customerio-node` you're using?

**Comment by rdrnt** [comment]
Hey @swastik, My `node` version is 10.21.0 & my `customerio-node` version is 3.0.2.

**Comment by swastik** [comment]
Thanks @rdrnt. I tried using `node` version `10.12.0` and `customerio-node` v3.0.2 but I wasn't able to reproduce this error.

If you have a stack trace, can you please share that? I'm trying to understand where the error is originating from.

**Comment by rdrnt** [comment]
@swastik Weird. Here's the stack trace:

```
>  Error tracking customerio TypeError [ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type Function. Received type object
>      at checkListener (events.js:55:11)
>      at ClientRequest.once (events.js:299:3)
>      at new ClientRequest (_http_client.js:160:10)
>      at new request (https.js:289:10)
>      at Object.obj.(anonymous function) (/Users/rileytiary/.config/yarn/global/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:248:31)
>      at Object.request (/Users/rileytiary/projects/blahblah/functions/node_modules/agent-base/patch-core.js:25:22)
>      at Promise (/Users/rileytiary/projects/blahblah/functions/node_modules/customerio-node/dist/lib/request.js:33:31)
>      at new Promise (<anonymous>)
>      at CIORequest.handler (/Users/rileytiary/projects/blahblah/functions/node_modules/customerio-node/dist/lib/request.js:31:16)
>      at CIORequest.post (/Users/rileytiary/projects/blahblah/functions/node_modules/customerio-node/dist/lib/request.js:74:21)
```

**Comment by swastik** [comment]
I believe this could be related to https://github.com/TooTallNate/node-agent-base/issues/24 and https://github.com/TooTallNate/node-agent-base/issues/29, however, I'm not sure what is using `agent-base` yet—it doesn't seem to be one of our dependencies. I'm checking a bit more.

**Comment by swastik** [comment]
@rdrnt Unfortunately I haven't been able to reproduce this at all. 😞

I have a feeling this might be related to some extensions or some package versions. Would you mind creating a Github repo with your node project that can help me reproduce? Feel free to remove any sensitive code.

**Comment by swastik** [comment]
@rdrnt Unfortunately I haven't been able to reproduce this at all. 😞

I have a feeling this might be related to some extensions or some package versions. Would you mind creating a Github repo with your node project that can help reproduce the issue? Feel free to remove any sensitive code.

**Comment by rdrnt** [comment]
@swastik The project is for firebase functions, so it's a bit hard to create an example repo/project. 

I ended up upgrading the project to node 14 and it it seems to have resolved the issue.

**Comment by swastik** [comment]
Glad to hear that, @rdrnt! 😄 I'm going to close this issue.

**Comment by benspring** [comment]
I'm having the same problem, and am on node v14 ~ Any other fix recommendations (my base OS is Linux Mint).

**Comment by swastik** [comment]
@benspring Is your project also for firebase functions? If not, do you mind creating a minimal Github repo with your project that can help reproduce the issue? I'll check out node v14 again as well.

**Comment by DanielJanke** [comment]
Got the same problem...

**Comment by swastik** [comment]
@DanielJanke We're still looking into this, trying to reproduce the issue. If you have a minimal Github repo with the issue, that will be immensely helpful!

**Comment by mike-engel** [comment]
Reopening until we can determine the exact cause.

**Comment by morleyzhi** [comment]
@mike-engel My team ran into this issue late last week!

The problem is you are calling `https.request` here with `uri, options, callback`:
https://github.com/customerio/customerio-node/blob/main/lib/request.ts#L63

But `https.request` only takes two params, `options, callback`:
https://node.readthedocs.io/en/stable/api/https/#httpsrequestoptions-callback

Thanks to @vcarl for finding this bug!

**Comment by mike-engel** [comment]
@morleyzhi Thanks for those links. Looking at the official `https` docs from the primary node website, there is a [signature for `https.request` available](https://nodejs.org/dist/latest-v16.x/docs/api/https.html#https_https_request_url_options_callback) since `10.9.0` to allow the `url` to be passed separately from the `options` object.

May I ask what version of node you're using? The documentation on Read The Docs looks to be out of date, incomplete, or ahead of the main docs.

**Comment by morleyzhi** [comment]
@mike-engel In our case, we're using the package in a React Native env, which uses a shim for http support. That shim only implements the one call signature:

https://github.com/tradle/react-native-http/blob/master/index.js#L14

**Comment by mike-engel** [comment]
I've created what I think will be a fix in #83. If you all would like to try it out, I've published it as `3.0.3-beta.1`. You can install it as follows:

```sh
# latest beta
npm i customerio-node@beta

# specific beta
npm i customerio-node@3.0.3-beta.1
```

**Comment by DanielJanke** [comment]
Woah! Thanks for the quick fix.

**Comment by mike-engel** [comment]
Just published as `3.0.3`. Thanks for the help, y'all!
